woocommerce/woocommerce-ios
# [Mobile Payments] Process a payment

Closes #3825, #3750, #3841

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Please note there are several parts of the implementation that are not production ready.⚠️

🔴🔴 Testing this PR requires a hardware reader and setting up a store for testing🔴🔴

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by implementing logic in Yosemite and Hardware to process a payment.

This is a two step process, on top of the previous step implemented in PR #3827 (creating a Payment Intent on device)

After creating a payment intent, we request collecting a payment method (which basically implies tapping the card, or inserting it into the card reader) and then process the payment.

## Changes
* Modified the CardReaderService API, to return a Payment Intent only when the payment has been processed, instead of at every step of the sequence (e.g. after collecting a payment method).
* Modified `collectPayment` in CardPresentPaymentStore to trigger the three operations (create intent, collect payment method and process payment) in sequence
* Add code to OrderDetailsViewModel and OrderDetailsViewController to trigger the payment capture process. For testing purposes, we added a button to the toolbar, to keep things simple.

There are several logs left in place. Some of them indicate functionality that we do have to implement, like:

```
==== did request reader input
SCPReaderInputOptions(rawValue: 7)
7
//// did request reader input
==== did request display message
5
//// did request display message
```

These tells us that the reader expects us to show user-facing messages asking users to tap / insert a card. There will be tickets added soon for these.

We also log the final payment intent, after the payment processing is completed:

```
==== Yosemite log for testing. Payment intent after processing payment
pi_REDACTED
//// payment intent processed 
```

## How to test
Testing is a bit complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

Once the store is setup and there is a hardware reader available:
1. Checkout the branch. Run `bundle exec pod install`. 
2. Build and run on a device. 
3. Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
4. Log in to an account that matches a store that has been configured for testing.
5. Navigate to Settings in the app. (Tap the gear button in the top right)
6. In the "Store Settings" section, tap "Manage Card Reader".
7. Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
8. Tap "Connect card reader"
9. Wait until the app suggests one reader to connect to.
10. Tap "Connect to reader"
11. Make sure the reader light turns solid blue.
12. Navigate to an order (any order).
13. Tap the money emojis in the top right corner
14. Wait until the reader beeps, one small square blue led turns on, and tap a card.
15. Wait until the reader beeps.
16. Wait until the UI presents an alert showing success or error. 

The following video hopefully helps illustrate the process:


https://user-images.githubusercontent.com/2722505/111704327-54016380-8815-11eb-9feb-05c78f3a9eae.mov


We tried to provide enough UI so that the test can be actually provide visual feedback, without going down the rabbit hole of fully implementing error handling

Update release notes:
